utopia-remix/app/handlers/validators.spec.ts

```diff
@@ -0,0 +1,211 @@
+import { prisma } from '../db.server'
+import {
+  createTestProject,
+  createTestProjectAccess,
+  createTestSession,
+  createTestUser,
+  newTestRequest,
+  truncateTables,
+} from '../test-util'
+import { validateProjectAccess } from './validators'
+import * as permissionsService from '../services/permissionsService.server'
+import { AccessLevel, UserProjectPermission } from '../types'
+import { ApiError } from '../util/errors'
+import type { ValidationResult } from '../util/api.server'
+
+describe('validators', () => {
+  describe('validateProjectAccess', () => {
+    let hasUserProjectPermission: jest.SpyInstance
+
+    afterEach(async () => {
+      await truncateTables([
+        prisma.projectID,
+        prisma.projectAccess,
+        prisma.project,
+        prisma.userDetails,
+        prisma.persistentSession,
+      ])
+      hasUserProjectPermission.mockClear()
+      jest.spyOn(permissionsService, 'hasUserProjectPermission').mockRestore()
+    })
+    beforeEach(async () => {
+      await createTestUser(prisma, { id: 'bob' })
+      await createTestUser(prisma, { id: 'alice' })
+      await createTestSession(prisma, { key: 'bob-key', userId: 'bob' })
+      await createTestSession(prisma, { key: 'alice-key', userId: 'alice' })
+      await createTestProject(prisma, { id: 'one', ownerId: 'bob' })
+
+      hasUserProjectPermission = jest.spyOn(permissionsService, 'hasUserProjectPermission')
```

Comment:
I ended up simplifying it much. TBH I'm not that familiar with the jest mocking so I just copy-pasted from https://github.com/concrete-utopia/utopia/blob/2372876aec02dbb1f524a5adcbfc1d9e2bff0790/utopia-remix/app/routes-test/internal.projects.%24id.access.spec.ts#L44, it probably would be good to fix it over there as well (it should be applicable I think)

Reply:
Yes, note that there the variable is just being set in one place, and is later being used in all of the tests (it's just a reference for the mock itself, instead of creating it every time)